package-name-map/src/test/package-name-map_test.ts

```diff
@@ -0,0 +1,281 @@
+            },
+            lodash: {
+              path: '/node_modules/lodash-es',
+              main: 'lodash.js',
+            },
+            relative: {
+              path: 'node_modules/relative',
+              main: 'relative.js',
+            },
+            '@polymer/polymer': {
+              path: '/node_modules/@polymer/polymer',
+              main: 'polymer.js',
+            },
+            '@polymer/polymer-foo': {
+              path: '/node_modules/@polymer/polymer-foo',
+              main: 'polymer-foo.js',
+            },
+          },
+        },
+        referrerURL
+      );
+
+      test('resolves package name for a package with only a main', () => {
+        assert.equal(
+          map.resolve('app', referrerURL),
+          'http://foo.com/bar/app/src/index.js'
+        );
+      });
+
+      test('resolves package name for a package with an absolute path and main', () => {
+        assert.equal(
+          map.resolve('lodash', referrerURL),
+          'http://foo.com/node_modules/lodash-es/lodash.js'
+        );
+      });
+
+      test('resolves a submodule for package with an absolute path and main', () => {
+        assert.equal(
+          map.resolve('lodash/bar.js', referrerURL),
+          'http://foo.com/node_modules/lodash-es/bar.js'
+        );
+      });
+
+      test('resolves package name for a package with a relative path and main', () => {
+        assert.equal(
+          map.resolve('relative', referrerURL),
+          'http://foo.com/bar/node_modules/relative/relative.js'
+        );
+      });
+
+      test('resolves a submodule for package with a relative path and main', () => {
+        assert.equal(
+          map.resolve('relative/bar.js', referrerURL),
+          'http://foo.com/bar/node_modules/relative/bar.js'
+        );
+      });
+
+      test('resolves a "scoped" package name', () => {
+        assert.equal(
+          map.resolve('@polymer/polymer', referrerURL),
+          'http://foo.com/node_modules/@polymer/polymer/polymer.js'
+        );
+      });
+
+      test('resolves a package name with another name as a prefix', () => {
+        assert.equal(
+          map.resolve('@polymer/polymer-foo', referrerURL),
+          'http://foo.com/node_modules/@polymer/polymer-foo/polymer-foo.js'
+        );
+      });
+
+      test('errors for a submodule with only a main', () => {
+        assert.throws(() => map.resolve('app/foo.js', referrerURL));
+      });
+    });
+
+    suite('path_prefix', () => {
+      const map = new PackageNameMap(
+        {
+          path_prefix: '/node_modules',
+          packages: {
+            moment: {
+              main: 'moment.js',
+            },
+            lodash: {
+              path: 'lodash-es',
+              main: 'lodash.js',
+            },
+            '@polymer/polymer': {
+              path: '@polymer/polymer',
+              main: 'polymer.js',
+            },
+          },
+        },
+        'http://foo.com/'
+      );
+
```

Comment:
How are percent encodings of package names handled? `"fs-thing"` vs. `"%66%73-thing"`?

Reply:
Good question. Can you file a separate issue to make sure we add test coverage and consider it carefully? My intention is that package names should be treated as opaque and not decoded or encoded in that regard.